x/slashing/hooks.go

```diff
@@ -0,0 +1,46 @@
+package slashing
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// Create a new slashing period when a validator is bonded
+func (k Keeper) onValidatorBonded(ctx sdk.Context, address sdk.ValAddress) {
+	slashingPeriod := ValidatorSlashingPeriod{
+		ValidatorAddr: address,
+		StartHeight:   ctx.BlockHeight(),
+		EndHeight:     0,
+		SlashedSoFar:  sdk.ZeroDec(),
+	}
+	k.setValidatorSlashingPeriod(ctx, slashingPeriod)
+}
+
+// Mark the slashing period as having ended when a validator is unbonded
+func (k Keeper) onValidatorUnbonded(ctx sdk.Context, address sdk.ValAddress) {
+	slashingPeriod := k.getValidatorSlashingPeriodForHeight(ctx, address, ctx.BlockHeight())
+	slashingPeriod.EndHeight = ctx.BlockHeight()
+	k.setValidatorSlashingPeriod(ctx, slashingPeriod)
+}
+
+// Wrapper struct for sdk.ValidatorHooks
+type ValidatorHooks struct {
+	k Keeper
+}
+
+// Return a sdk.ValidatorHooks interface over the wrapper struct
+func (k Keeper) ValidatorHooks() sdk.ValidatorHooks {
```

Comment:
This is to provide stronger object-capabilities security, since you can't access the underlying `slashing.Keeper` struct. Open to discussion but probably should be in the context of https://github.com/cosmos/cosmos-sdk/issues/802#issuecomment-380391802.

Reply:
Oh I think I understand - we want to wrap another object which has an unexposed field for protection against reflecting the interface to get at the underlying struct, correct? - if so this thread should be marked as resolved